File: Common/Utilities.Text.vb
Imports System.Runtime.CompilerServices
Imports System.Text
Imports Media = System.Windows.Media
Imports Reflect = System.Reflection
Imports SIO = System.IO

Namespace Common

    Partial Class Utilities

        Public Class Text

            Private Sub New()
                Throw New NotSupportedException()
            End Sub

#Region "Escape Support Logic"
            Private Shared ReadOnly EscapeDic As Dictionary(Of Char, Char) =
                (Function() New Dictionary(Of Char, Char) From {
                         {"0"c, Strings.ChrW(&H0)},
                         {""""c, Strings.ChrW(&H27)},
                         {"'"c, Strings.ChrW(&H22)},
                         {"?"c, Strings.ChrW(&H3F)},
                         {"\"c, Strings.ChrW(&H5C)},
                         {"a"c, Strings.ChrW(&H7)},
                         {"b"c, Strings.ChrW(&H8)},
                         {"f"c, Strings.ChrW(&HC)},
                         {"n"c, Strings.ChrW(&HA)},
                         {"r"c, Strings.ChrW(&HD)},
                         {"t"c, Strings.ChrW(&H9)},
                         {"v"c, Strings.ChrW(&HB)}
                     }).Invoke()

            Public Shared Function IsBinaryDigit(ByVal C As Char) As Boolean
                Return C = "0"c OrElse C = "1"c
            End Function

            Public Shared Function IsDecimalDigit(ByVal C As Char) As Boolean
                Return "0"c <= C AndAlso C <= "9"c
            End Function

            Public Shared Function IsOctalDigit(ByVal C As Char) As Boolean
                Return "0"c <= C AndAlso C <= "7"c
            End Function

            Public Shared Function IsHexadecimalDigit(ByVal C As Char) As Boolean
                Return ("0"c <= C AndAlso C <= "9"c) OrElse ("A"c <= C AndAlso C <= "F"c) OrElse ("a"c <= C AndAlso C <= "f"c)
            End Function

            Public Shared Function CEscapeC(ByVal Input As String, Optional ByVal DoesThrow As Boolean = True) As Char
                Dim T1, T2 As Char

                If Input.Length = 0 Then
                    Throw New ArgumentException()
                End If

                T2 = Input.Chars(0)

                If T2 <> "\" Then
                    If DoesThrow AndAlso Input.Length <> 1 Then
                        Throw New ArgumentException("Invalid escaped character.")
                    End If
                    Return T2
                End If

                T1 = Input.Chars(1)

                If EscapeDic.TryGetValue(T1, T2) Then
                    If DoesThrow AndAlso Input.Length <> 2 Then
                    End If
                    Return T2
                End If

                If T1 = "x"c Then
                    If Input.Length <> 4 Then
                        If DoesThrow Then
                            Throw New ArgumentException("Invalid escaped string.")
                        Else
                            Return T1
                        End If
                    End If
                    If Not (IsHexadecimalDigit(Input.Chars(2)) AndAlso
                            IsHexadecimalDigit(Input.Chars(3))) Then
                        If DoesThrow Then
                            Throw New ArgumentException("Invalid escaped string.")
                        Else
                            Return T1
                        End If
                    End If

                    Return Strings.ChrW(Convert.ToInt32(Input.Substring(2, 2), 16))
                End If

                If T1 = "u"c Then
                    If Input.Length <> 6 Then
                        If DoesThrow Then
                            Throw New ArgumentException("Invalid escaped string.")
                        Else
                            Return T1
                        End If
                    End If
                    If Not (IsHexadecimalDigit(Input.Chars(2)) AndAlso
                            IsHexadecimalDigit(Input.Chars(3)) AndAlso
                            IsHexadecimalDigit(Input.Chars(4)) AndAlso
                            IsHexadecimalDigit(Input.Chars(5))) Then
                        If DoesThrow Then
                            Throw New ArgumentException("Invalid escaped string.")
                        Else
                            Return T1
                        End If
                    End If

                    Return Strings.ChrW(Convert.ToInt32(Input.Substring(2, 4), 16))
                End If

                If T1 = "U"c Then
                    If Input.Length <> 10 Then
                        If DoesThrow Then
                            Throw New ArgumentException("Invalid escaped string.")
                        Else
                            Return T1
                        End If
                    End If
                    If Not (IsHexadecimalDigit(Input.Chars(2)) AndAlso
                            IsHexadecimalDigit(Input.Chars(3)) AndAlso
                            IsHexadecimalDigit(Input.Chars(4)) AndAlso
                            IsHexadecimalDigit(Input.Chars(5)) AndAlso
                            IsHexadecimalDigit(Input.Chars(6)) AndAlso
                            IsHexadecimalDigit(Input.Chars(7)) AndAlso
                            IsHexadecimalDigit(Input.Chars(8)) AndAlso
                            IsHexadecimalDigit(Input.Chars(9))) Then
                        If DoesThrow Then
                            Throw New ArgumentException("Invalid escaped string.")
                        Else
                            Return T1
                        End If
                    End If

                    Return Strings.ChrW(Convert.ToInt32(Input.Substring(2, 8), 16))
                End If

                If IsOctalDigit(T1) Then
                    If Input.Length <> 4 Then
                        If DoesThrow Then
                            Throw New ArgumentException("Invalid escaped string.")
                        Else
                            Return T1
                        End If
                    End If
                    If Not (IsOctalDigit(Input.Chars(2)) AndAlso
                            IsOctalDigit(Input.Chars(3))) Then
                        If DoesThrow Then
                            Throw New ArgumentException("Invalid escaped string.")
                        Else
                            Return T1
                        End If
                    End If

                    Return Strings.ChrW(Convert.ToInt32(Input.Substring(1, 3), 8))
                End If

                If DoesThrow Then
                    Throw New ArgumentException("Invalid escaped string.")
                Else
                    Return T1
                End If
            End Function

            Public Shared Function CEscape(ByVal Input As String, Optional ByVal DoesThrow As Boolean = True) As String
                Dim T1, T2 As Char

                Dim Res = New StringBuilder()

                For I As Integer = 0 To Input.Length - 1
                    T2 = Input.Chars(I)

                    If T2 = "\"c Then
                        I += 1

                        If I = Input.Length Then
                            If DoesThrow Then
                                Throw New ArgumentException("Invalid escaped string.")
                            Else
                                Exit For
                            End If
                        End If

                        T1 = Input.Chars(I)

                        If EscapeDic.TryGetValue(T1, T2) Then
                            Res.Append(T2)
                            Continue For
                        End If

                        If T1 = "x"c Then
                            If I + 2 >= Input.Length Then
                                If DoesThrow Then
                                    Throw New ArgumentException("Invalid escaped string.")
                                Else
                                    Res.Append(T1)
                                    Continue For
                                End If
                            End If
                            If Not (IsHexadecimalDigit(Input.Chars(I + 1)) AndAlso
                                    IsHexadecimalDigit(Input.Chars(I + 2))) Then
                                If DoesThrow Then
                                    Throw New ArgumentException("Invalid escaped string.")
                                Else
                                    Res.Append(T1)
                                    Continue For
                                End If
                            End If

                            Res.Append(Strings.ChrW(Convert.ToInt32(Input.Substring(I + 1, 2), 16)))
                            I += 2
                            Continue For
                        End If

                        If T1 = "u"c Then
                            If I + 4 >= Input.Length Then
                                If DoesThrow Then
                                    Throw New ArgumentException("Invalid escaped string.")
                                Else
                                    Res.Append(T1)
                                    Continue For
                                End If
                            End If
                            If Not (IsHexadecimalDigit(Input.Chars(I + 1)) AndAlso
                                    IsHexadecimalDigit(Input.Chars(I + 2)) AndAlso
                                    IsHexadecimalDigit(Input.Chars(I + 3)) AndAlso
                                    IsHexadecimalDigit(Input.Chars(I + 4))) Then
                                If DoesThrow Then
                                    Throw New ArgumentException("Invalid escaped string.")
                                Else
                                    Res.Append(T1)
                                    Continue For
                                End If
                            End If

                            Res.Append(Strings.ChrW(Convert.ToInt32(Input.Substring(I + 1, 4), 16)))
                            I += 4
                            Continue For
                        End If

                        If T1 = "U"c Then
                            If I + 8 >= Input.Length Then
                                If DoesThrow Then
                                    Throw New ArgumentException("Invalid escaped string.")
                                Else
                                    Res.Append(T1)
                                    Continue For
                                End If
                            End If
                            If Not (IsHexadecimalDigit(Input.Chars(I + 1)) AndAlso
                                    IsHexadecimalDigit(Input.Chars(I + 2)) AndAlso
                                    IsHexadecimalDigit(Input.Chars(I + 3)) AndAlso
                                    IsHexadecimalDigit(Input.Chars(I + 4)) AndAlso
                                    IsHexadecimalDigit(Input.Chars(I + 5)) AndAlso
                                    IsHexadecimalDigit(Input.Chars(I + 6)) AndAlso
                                    IsHexadecimalDigit(Input.Chars(I + 7)) AndAlso
                                    IsHexadecimalDigit(Input.Chars(I + 8))) Then
                                If DoesThrow Then
                                    Throw New ArgumentException("Invalid escaped string.")
                                Else
                                    Res.Append(T1)
                                    Continue For
                                End If
                            End If

                            Res.Append(Strings.ChrW(Convert.ToInt32(Input.Substring(I + 1, 8), 16)))
                            I += 8
                            Continue For
                        End If

                        If IsOctalDigit(T1) Then
                            If I + 2 >= Input.Length Then
                                If DoesThrow Then
                                    Throw New ArgumentException("Invalid escaped string.")
                                Else
                                    Res.Append(T1)
                                    Continue For
                                End If
                            End If
                            If Not (IsOctalDigit(Input.Chars(I + 1)) AndAlso
                                    IsOctalDigit(Input.Chars(I + 2))) Then
                                If DoesThrow Then
                                    Throw New ArgumentException("Invalid escaped string.")
                                Else
                                    Res.Append(T1)
                                    Continue For
                                End If
                            End If

                            Res.Append(Strings.ChrW(Convert.ToInt32(Input.Substring(I, 3), 8)))
                            I += 2
                            Continue For
                        End If

                        If DoesThrow Then
                            Throw New ArgumentException("Invalid escaped string.")
                        Else
                            Res.Append(T1)
                            Continue For
                        End If
                    End If
                    Res.Append(T2)
                Next

                Return Res.ToString()
            End Function
#End Region

            Public Shared Function EscapeNewLine(ByVal Str As String) As String
                Dim Res = New StringBuilder()

                For Each Ch In Str
                    Select Case Ch
                        Case ControlChars.Cr
                            Res.Append("\r")
                        Case ControlChars.Lf
                            Res.Append("\n")
                        Case "\"c
                            Res.Append("\\")
                        Case Else
                            Res.Append(Ch)
                    End Select
                Next

                Return Res.ToString()
            End Function

            Public Shared Function UnescapeNewLine(ByVal Str As String) As String
                Dim Res = New StringBuilder()

                For I As Integer = 0 To Str.Length - 1
                    Dim Ch = Str.Chars(I)

                    If Ch = "\"c Then
                        I += 1
                        If I = Str.Length Then
                            Throw New Exception("Invalid list string.")
                        End If
                        Ch = Str.Chars(I)

                        Select Case Ch
                            Case "r"c
                                Res.Append(ControlChars.Cr)
                            Case "n"c
                                Res.Append(ControlChars.Lf)
                            Case "\"c
                                Res.Append("\")
                            Case Else
                                Throw New Exception("Invalid escape character.")
                        End Select
                    End If

                    Res.Append(Ch)
                Next

                Return Res.ToString()
            End Function

            Public Shared ReadOnly Property CurruntFormatProvider As IFormatProvider
                Get
                    Return Nothing
                End Get
            End Property

            Public Shared Function FirstCapitalized(ByVal Str As String) As String
                Return Char.ToUpper(Str.Chars(0)) & Str.Substring(1).ToLower()
            End Function

            Public Shared Function EnumerableToString(Of T)(ByVal Enumerable As IEnumerable(Of T)) As String
                Dim Res = New StringBuilder("{"c)
                Dim Bl = True

                For Each I As T In Enumerable
                    If Bl Then
                        Bl = False
                    Else
                        Res.Append(", ")
                    End If
                    Res.Append(I)
                Next

                Return Res.Append("}"c).ToString()
            End Function

            Public Shared Function PadAlignString(Str As String, Ch As Char, Length As Integer, Alignment As TextAlignment) As String
                Verify.False(Length < Str.Length, "Length of Str is less than Length.")

                Dim N = Str.Length

                Dim Res = ""
                Select Case Alignment
                    Case TextAlignment.Left
                        Res = Str.PadRight(Length, Ch)
                    Case TextAlignment.Center
                        Res = Str.PadRight(Length - N \ 2, Ch).PadLeft(Length, Ch)
                    Case TextAlignment.Right
                        Res = Str.PadLeft(Length, Ch)
                    Case Else
                        Verify.FailArg(NameOf(Alignment), "Invalid Alignment.")
                End Select

                Return Res
            End Function

            Public Enum TextAlignment
                Right
                Center
                Left
            End Enum

        End Class

    End Class

End Namespace

File: Common/Utilities.Text.cs
using Microsoft.VisualBasic;
using System.Collections.Generic;
using System;
using System.Text;
using Microsoft.VisualBasic.CompilerServices;

namespace Ks
{
    namespace Common
    {
        partial class Utilities
        {
            public class Text
            {
                private Text()
                {
                    throw new NotSupportedException();
                }

                private static readonly Dictionary<char, char> EscapeDic = (() => new Dictionary<char, char>()
                {
                    {
                        '0',
                        (char)0x0
                    },
                    {
                        '"',
                        (char)0x27
                    },
                    {
                        '\'',
                        (char)0x22
                    },
                    {
                        '?',
                        (char)0x3F
                    },
                    {
                        '\\',
                        (char)0x5C
                    },
                    {
                        'a',
                        (char)0x7
                    },
                    {
                        'b',
                        (char)0x8
                    },
                    {
                        'f',
                        (char)0xC
                    },
                    {
                        'n',
                        (char)0xA
                    },
                    {
                        'r',
                        (char)0xD
                    },
                    {
                        't',
                        (char)0x9
                    },
                    {
                        'v',
                        (char)0xB
                    }
                }).Invoke();

                public static bool IsBinaryDigit(char C)
                {
                    return C == '0' || C == '1';
                }

                public static bool IsDecimalDigit(char C)
                {
                    return '0' <= C && C <= '9';
                }

                public static bool IsOctalDigit(char C)
                {
                    return '0' <= C && C <= '7';
                }

                public static bool IsHexadecimalDigit(char C)
                {
                    return ('0' <= C && C <= '9') || ('A' <= C && C <= 'F') || ('a' <= C && C <= 'f');
                }

                public static char CEscapeC(string Input, bool DoesThrow = true)
                {
                    char T1, T2;

                    if (Input.Length == 0)
                        throw new ArgumentException();

                    T2 = Input.Chars[0];

                    if (Conversions.ToString(T2) != @"\")
                    {
                        if (DoesThrow && Input.Length != 1)
                            throw new ArgumentException("Invalid escaped character.");
                        return T2;
                    }

                    T1 = Input.Chars[1];

                    if (EscapeDic.TryGetValue(T1, out T2))
                    {
                        if (DoesThrow && Input.Length != 2)
                        {
                        }
                        return T2;
                    }

                    if (T1 == 'x')
                    {
                        if (Input.Length != 4)
                        {
                            if (DoesThrow)
                                throw new ArgumentException("Invalid escaped string.");
                            else
                                return T1;
                        }
                        if (!(IsHexadecimalDigit(Input.Chars[2]) && IsHexadecimalDigit(Input.Chars[3])))
                        {
                            if (DoesThrow)
                                throw new ArgumentException("Invalid escaped string.");
                            else
                                return T1;
                        }

                        return (char)Convert.ToInt32(Input.Substring(2, 2), 16);
                    }

                    if (T1 == 'u')
                    {
                        if (Input.Length != 6)
                        {
                            if (DoesThrow)
                                throw new ArgumentException("Invalid escaped string.");
                            else
                                return T1;
                        }
                        if (!(IsHexadecimalDigit(Input.Chars[2]) && IsHexadecimalDigit(Input.Chars[3]) && IsHexadecimalDigit(Input.Chars[4]) && IsHexadecimalDigit(Input.Chars[5])))
                        {
                            if (DoesThrow)
                                throw new ArgumentException("Invalid escaped string.");
                            else
                                return T1;
                        }

                        return (char)Convert.ToInt32(Input.Substring(2, 4), 16);
                    }

                    if (T1 == 'U')
                    {
                        if (Input.Length != 10)
                        {
                            if (DoesThrow)
                                throw new ArgumentException("Invalid escaped string.");
                            else
                                return T1;
                        }
                        if (!(IsHexadecimalDigit(Input.Chars[2]) && IsHexadecimalDigit(Input.Chars[3]) && IsHexadecimalDigit(Input.Chars[4]) && IsHexadecimalDigit(Input.Chars[5]) && IsHexadecimalDigit(Input.Chars[6]) && IsHexadecimalDigit(Input.Chars[7]) && IsHexadecimalDigit(Input.Chars[8]) && IsHexadecimalDigit(Input.Chars[9])))
                        {
                            if (DoesThrow)
                                throw new ArgumentException("Invalid escaped string.");
                            else
                                return T1;
                        }

                        return (char)Convert.ToInt32(Input.Substring(2, 8), 16);
                    }

                    if (IsOctalDigit(T1))
                    {
                        if (Input.Length != 4)
                        {
                            if (DoesThrow)
                                throw new ArgumentException("Invalid escaped string.");
                            else
                                return T1;
                        }
                        if (!(IsOctalDigit(Input.Chars[2]) && IsOctalDigit(Input.Chars[3])))
                        {
                            if (DoesThrow)
                                throw new ArgumentException("Invalid escaped string.");
                            else
                                return T1;
                        }

                        return (char)Convert.ToInt32(Input.Substring(1, 3), 8);
                    }

                    if (DoesThrow)
                        throw new ArgumentException("Invalid escaped string.");
                    else
                        return T1;
                }

                public static string CEscape(string Input, bool DoesThrow = true)
                {
                    char T1 = default(char), T2 = default(char);

                    var Res = new StringBuilder();
                    var loopTo = Input.Length - 1;
                    for (int I = 0; I <= loopTo; I++)
                    {
                        T2 = Input.Chars[I];

                        if (T2 == '\\')
                        {
                            I += 1;

                            if (I == Input.Length)
                            {
                                if (DoesThrow)
                                    throw new ArgumentException("Invalid escaped string.");
                                else
                                    break;
                            }

                            T1 = Input.Chars[I];

                            if (EscapeDic.TryGetValue(T1, out T2))
                            {
                                Res.Append(T2);
                                continue;
                            }

                            if (T1 == 'x')
                            {
                                if ((I + 2) >= Input.Length)
                                {
                                    if (DoesThrow)
                                        throw new ArgumentException("Invalid escaped string.");
                                    else
                                    {
                                        Res.Append(T1);
                                        continue;
                                    }
                                }
                                if (!(IsHexadecimalDigit(Input.Chars[I + 1]) && IsHexadecimalDigit(Input.Chars[I + 2])))
                                {
                                    if (DoesThrow)
                                        throw new ArgumentException("Invalid escaped string.");
                                    else
                                    {
                                        Res.Append(T1);
                                        continue;
                                    }
                                }

                                Res.Append((char)Convert.ToInt32(Input.Substring(I + 1, 2), 16));
                                I += 2;
                                continue;
                            }

                            if (T1 == 'u')
                            {
                                if ((I + 4) >= Input.Length)
                                {
                                    if (DoesThrow)
                                        throw new ArgumentException("Invalid escaped string.");
                                    else
                                    {
                                        Res.Append(T1);
                                        continue;
                                    }
                                }
                                if (!(IsHexadecimalDigit(Input.Chars[I + 1]) && IsHexadecimalDigit(Input.Chars[I + 2]) && IsHexadecimalDigit(Input.Chars[I + 3]) && IsHexadecimalDigit(Input.Chars[I + 4])))
                                {
                                    if (DoesThrow)
                                        throw new ArgumentException("Invalid escaped string.");
                                    else
                                    {
                                        Res.Append(T1);
                                        continue;
                                    }
                                }

                                Res.Append((char)Convert.ToInt32(Input.Substring(I + 1, 4), 16));
                                I += 4;
                                continue;
                            }

                            if (T1 == 'U')
                            {
                                if ((I + 8) >= Input.Length)
                                {
                                    if (DoesThrow)
                                        throw new ArgumentException("Invalid escaped string.");
                                    else
                                    {
                                        Res.Append(T1);
                                        continue;
                                    }
                                }
                                if (!(IsHexadecimalDigit(Input.Chars[I + 1]) && IsHexadecimalDigit(Input.Chars[I + 2]) && IsHexadecimalDigit(Input.Chars[I + 3]) && IsHexadecimalDigit(Input.Chars[I + 4]) && IsHexadecimalDigit(Input.Chars[I + 5]) && IsHexadecimalDigit(Input.Chars[I + 6]) && IsHexadecimalDigit(Input.Chars[I + 7]) && IsHexadecimalDigit(Input.Chars[I + 8])))
                                {
                                    if (DoesThrow)
                                        throw new ArgumentException("Invalid escaped string.");
                                    else
                                    {
                                        Res.Append(T1);
                                        continue;
                                    }
                                }

                                Res.Append((char)Convert.ToInt32(Input.Substring(I + 1, 8), 16));
                                I += 8;
                                continue;
                            }

                            if (IsOctalDigit(T1))
                            {
                                if ((I + 2) >= Input.Length)
                                {
                                    if (DoesThrow)
                                        throw new ArgumentException("Invalid escaped string.");
                                    else
                                    {
                                        Res.Append(T1);
                                        continue;
                                    }
                                }
                                if (!(IsOctalDigit(Input.Chars[I + 1]) && IsOctalDigit(Input.Chars[I + 2])))
                                {
                                    if (DoesThrow)
                                        throw new ArgumentException("Invalid escaped string.");
                                    else
                                    {
                                        Res.Append(T1);
                                        continue;
                                    }
                                }

                                Res.Append((char)Convert.ToInt32(Input.Substring(I, 3), 8));
                                I += 2;
                                continue;
                            }

                            if (DoesThrow)
                                throw new ArgumentException("Invalid escaped string.");
                            else
                            {
                                Res.Append(T1);
                                continue;
                            }
                        }
                        Res.Append(T2);
                    }

                    return Res.ToString();
                }

                public static string EscapeNewLine(string Str)
                {
                    var Res = new StringBuilder();

                    foreach (var Ch in Str)
                    {
                        switch (Ch)
                        {
                            case ControlChars.Cr:
                                {
                                    Res.Append(@"\r");
                                    break;
                                }

                            case ControlChars.Lf:
                                {
                                    Res.Append(@"\n");
                                    break;
                                }

                            case '\\':
                                {
                                    Res.Append(@"\\");
                                    break;
                                }

                            default:
                                {
                                    Res.Append(Ch);
                                    break;
                                }
                        }
                    }

                    return Res.ToString();
                }

                public static string UnescapeNewLine(string Str)
                {
                    var Res = new StringBuilder();
                    var loopTo = Str.Length - 1;
                    for (int I = 0; I <= loopTo; I++)
                    {
                        var Ch = Str.Chars[I];

                        if (Ch == '\\')
                        {
                            I += 1;
                            if (I == Str.Length)
                                throw new Exception("Invalid list string.");
                            Ch = Str.Chars[I];

                            switch (Ch)
                            {
                                case 'r':
                                    {
                                        Res.Append(ControlChars.Cr);
                                        break;
                                    }

                                case 'n':
                                    {
                                        Res.Append(ControlChars.Lf);
                                        break;
                                    }

                                case '\\':
                                    {
                                        Res.Append(@"\");
                                        break;
                                    }

                                default:
                                    {
                                        throw new Exception("Invalid escape character.");
                                        break;
                                    }
                            }
                        }

                        Res.Append(Ch);
                    }

                    return Res.ToString();
                }

                public static IFormatProvider CurruntFormatProvider
                {
                    get
                    {
                        return null;
                    }
                }

                public static string FirstCapitalized(string Str)
                {
                    return Conversions.ToString(char.ToUpper(Str.Chars[0])) + Str.Substring(1).ToLower();
                }

                public static string EnumerableToString<T>(IEnumerable<T> Enumerable)
                {
                    var Res = new StringBuilder(Conversions.ToString('{'));
                    var Bl = true;

                    foreach (T I in Enumerable)
                    {
                        if (Bl)
                            Bl = false;
                        else
                            Res.Append(", ");
                        Res.Append(I);
                    }

                    return Res.Append('}').ToString();
                }

                public static string PadAlignString(string Str, char Ch, int Length, TextAlignment Alignment)
                {
                    Verify.False(Length < Str.Length, "Length of Str is less than Length.");

                    var N = Str.Length;

                    var Res = "";
                    switch (Alignment)
                    {
                        case TextAlignment.Left:
                            {
                                Res = Str.PadRight(Length, Ch);
                                break;
                            }

                        case TextAlignment.Center:
                            {
                                Res = Str.PadRight(Length - (N / 2), Ch).PadLeft(Length, Ch);
                                break;
                            }

                        case TextAlignment.Right:
                            {
                                Res = Str.PadLeft(Length, Ch);
                                break;
                            }

                        default:
                            {
                                Verify.FailArg(nameof(Alignment), "Invalid Alignment.");
                                break;
                            }
                    }

                    return Res;
                }

                public enum TextAlignment
                {
                    Right,
                    Center,
                    Left
                }
            }
        }
    }
}
